### ISSUE TITLE
Support for cross-namespace references / GatewayAPI ReferenceGrants

### ISSUE THREAD

**Issue by pascal-hofmann** [traefik/traefik#10338]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

Traefiks Gateway API support currently does not include cross-name references via [ReferenceGrants](https://gateway-api.sigs.k8s.io/api-types/referencegrant/).

I could work on a PR of this (at least for cross-namespace references for secrets). Is this something that you would accept as PR?

(I would base my PR on #10205.)

**Comment by pascal-hofmann** [comment]
@mmatur I just pushed my current state. See https://github.com/mmatur/traefik/compare/feat/upgrade-gateway-api...pascal-hofmann:traefik:feat/reference-grants-for-secrets

I did not test this in a real setup and there are no integration tests. I just wrote unit tests and made them green.

**Comment by mmatur** [comment]
Hi @pascal-hofmann,

It could be a great addition to Traefik Gateway API implementation. 

I think you can create a PR for that.

**Comment by pascal-hofmann** [comment]
Awesome, I'll continue to work on this then.

@mmatur Is there an easy way to run the [Gateway API conformance tests](https://github.com/kubernetes-sigs/gateway-api/tree/4b1a28d0c874a3816d344fb0d4e2c04424294d54/conformance) against Traefik?

**Comment by mmatur** [comment]
Hi @pascal-hofmann,

Right now there is no easy way to run Gateway api conformance tests with Traefik. We have reworked integration tests on Traefik see PR #10334 it will help us to integrate gateway API conformance tests

I will open a PR in the following days to add them.

**Comment by pascal-hofmann** [comment]
Thanks.

For local testing I've now scripted:
- setting up an image registry and k8s cluster with k3d
- pushing the images and 
- running conformance tests.

The main thing that prevents them from running is the missing ObservedGeneration attribute for the GatewayClass status update. Once this is added, the tests run (but fail because of non-conformance).

```golang
--- a/pkg/provider/kubernetes/gateway/kubernetes.go
+++ b/pkg/provider/kubernetes/gateway/kubernetes.go
@@ -230,6 +232,7 @@ func (p *Provider) loadConfigurationFromGateway(ctx context.Context, client Clie
                        err := client.UpdateGatewayClassStatus(gatewayClass, metav1.Condition{
                                Type:               string(gatev1.GatewayClassConditionStatusAccepted),
                                Status:             metav1.ConditionTrue,
+                               ObservedGeneration: gatewayClass.Generation,
                                Reason:             "Handled",
                                Message:            "Handled by Traefik controller",
                                LastTransitionTime: metav1.Now(),
```

**Comment by pascal-hofmann** [comment]
@mmatur I found some more basic things. I'll create a separate PR against your gateway API 1.0 branch for these.

**Comment by mmatur** [comment]
@pascal-hofmann great, could you create a PR against traefik v3.0 branch?

**Comment by pascal-hofmann** [comment]
@mmatur I based all my work on the PR for GatewayAPI 1.0 and also used the conformance tests for this version, as they have additional tests for ReferenceGrants.

I’m not sure backporting to an older version of Gateway API makes sense. This would mean additional work now, and when merging #10205.

**Comment by mmatur** [comment]
Closed by #10346
